### ISSUE TITLE
Possibility to increase thread count & change DOKAN_OPTION_ALLOW_IPC_BATCHING? (FUSE)

### ISSUE THREAD

**Issue by algj** [dokan-dev/dokany#1202]
# Environment
 * Windows version: Win10
 * Processor architecture: x64, 2 threads
 * Dokany version: 2.1.0
 * Library type (Dokany/FUSE): FUSE 
 
# Check List
  - [X] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [X] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [X] I can always reproduce the issue with the provided description below.
  - [X] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [ ] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

# Description

I have been experimenting with the FUSE wrapper, but I got stuck with 2 threads. Is there a way to increase the amount of threads or spawn more threads without recompiling Dokany? On Linux, FUSE is a lot faster when dealing with multiple high latency reads, because it doesn't block IO, that's why I think min thread count should be increased, or at least customizable.

> CHANGELOG: `DOKAN_OPTIONS.ThreadCount` was replaced by `DOKAN_OPTIONS.SingleThread` since the library now uses a thread pool that allocates workers depending on workload and the available resources.

I assume this was possible before, but now it depends on the CPU thread count:

https://github.com/dokan-dev/dokany/blob/415ac366eaa7a257ceb217e79dab9f37fadc7d33/dokan/dokan.c#L791-L798

I'd suggest adding a way of changing thread count in Dokany & FUSE.

**Comment by Liryna** [comment]
Hi @algj ,
Have you tried [DOKAN_OPTION_ALLOW_IPC_BATCHING](https://github.com/dokan-dev/dokany/blob/master/dokan/dokan.h#L128-L133) ? There is no FUSE option (yet) but if you could force it and see if it helps your case.
Increasing main pull thread count is a limited solution that does not scale well with increase of IO activity.

**Comment by algj** [comment]
Oh! Thank you @Liryna! This is exactly what I was looking for! Unfortunately (from my understanding) you cannot change it without recompiling Dokany. There should be a way to change this option somehow.

I hope someone makes a PR regarding changing this option easily, e.g. setting preprocessor definition in CMake. *I could make a PR, but I'm sure it wouldn't be up to the standards.*

**Comment by LTRData** [comment]
No, the `DOKAN_OPTIONS_*` flags can be set in `Options` field in `DOKAN_OPTIONS` structure when mounting a new file system. It does not need recompiling driver or library.

**Comment by Liryna** [comment]
The problem is that those options are not directly available for FUSE, they need to be wrapped for the FUSE interface. See [here for DOKAN_OPTION_MOUNT_MANAGER](
https://github.com/dokan-dev/dokany/commit/c09d5ec5159aa07a4564d6dd9e20ad08f57648fe)

**Comment by LTRData** [comment]
> I'm a little confused... I don't mind the source code having both Dokany & FUSE code, but how does one do that?

The problem here was apparently that if you create a native Dokany implementation, this is very easy to set from your implementation by setting a flag in the options structure when mounting a new file system.

If on the other hand you implement a file system using the Fuse emulation layer on top of Dokany, this option is hidden in the emulation layer and not exposed to your implementation.

**Comment by algj** [comment]
Oh my!!! Thank you very much @Liryna for the PR!! ❤️ I thought I'll have to do the PR myself 😅

**Comment by algj** [comment]
I have tried out IPC_BATCHING. There is a clear performance difference when I did my tests (it's slower by a few ms consistently), but for some reason it still sticks to 2 threads...? I'm not sure if I'm misunderstanding something, I thought it's supposed to spawn new threads when there's a few of concurrent reads instead of blocking it all.

**Comment by algj** [comment]
Also, I think that ThreadCount should be added back (and possibly SingleThread option removed, but that may break some things?)

ThreadCount = 0, library decides what thread count to use
ThreadCount = 1, single threaded
ThreadCount >= 2, multi thread, user defines how many threads application should use

**Comment by Liryna** [comment]
Looks like the IPC_BATCHING option was not correctly set in the library https://github.com/dokan-dev/dokany/commit/aef92bcf23c0dea150e7864a4ef81984325fd6a5 :(
(The CI tests seems to fail, I will need to look at that https://ci.appveyor.com/project/Maxhy/dokany/builds/49214019 )

Regarding `ThreadCount`, I don't think that's a good idea as like I said before, it does not scale correctly. Right now the best performance-wise thread count is allocated and for slow IO we need to use the thread pool.

**Comment by cowboy823** [comment]
Hello Liryna.
I have same issue with algj.
I'm developing a web drive using memfs and ssh/ftp. This allows users to use some directories on the FTP server as if they were on a local drive.
In this case I'm doing multi-thread process since there is the transmission delay(or latency) between the server and user sides in the network .
So I have resolved the CreateFile, FindFile, WriteFile functions in multi-threading technical but not ReadFile function.
Dokan seems to handle the read requests in serial(i.e. single thread).
As above, I compile the dokan to have the mainPullThreadCount = 16 and tested read operation but no clear performance difference and seems to assign a thread per a file. So I can't speed up read operation(downloading).
How can I solve this problem?

**Comment by Liryna** [comment]
Note: This issue only affects computers with limited number of threads as dokan set the optimal number of threads based on the hardware.

If the application (notepad, etc) that issue Create / Readfile sequentially, dokan will receive them as it. If the app issue multiple async or multi-thread reads, dokan will send multiple reads request to memfs.
Your issue might be that each read are slow to execute due to network access ? Then you will need cache and precaching to improve the access of the data.

**Comment by cowboy823** [comment]
Thanks for your reply.
I'm going to process the read for a file in multi-thread. this is possible in dokan?

**Comment by Liryna** [comment]
Yes, memfs or the filesystem implementation can do anything it wants

**Comment by algj** [comment]
Currently FUSE wrapper still has problems of reading from different processes different files without blocking IO.

About the notepad example - I tried that out by making FUSE code sleep for 5 seconds before responding. Linux? Fine, opened both txt files in 5 seconds. Windows? Opened two different txt files in 10 seconds (assuming single CPU thread).

Not a problem if reading from local resources, but a big problem if reading from network where ping may reach 100ms, that could unnecessarily bottleneck everything if we're reading tons of small files - realistically we can't cache them.

[Linux solves this problem in two different ways:](https://fuse-devel.narkive.com/CodQceXu/how-could-fuse-support-parallel-opreation)

> 1) multi threaded
> 
> Libfuse starts (or reuses an idle one) a new thread for every request
> and multiple requests can be processed in parallel, one per thread.
> There is a limit on the number of pending requests and some operations
> block others. The former can be changed iirc while the later is required
> for correct behaviour and nothing can be done there.
> 
> 2) asyncron
> 
> Libfuse doesn't have a read-to-use main loop for this. But you can
> include the fuse FD in your select/poll/epoll loop and call the recieve
> and process functions in libfuse yourself. You can read requests from
> the FD and reply to them in any order. The kernel doesn't care. The
> limit on the number of pending requests and locking between some
> operation remain.

I suggest allowing to set thread count, since spawning threads may be a bit too much to implement.
